### PR TITLE
2.2

### PR DESCRIPTION
**Pull request by MihaiAnton**
Implements Part II.
Still work in progress.

- [x] isosurface raycasting
- [x] phong shading
- [x] bisection accuracy
- [x] fix phong shading issue

### CODE REVIEW

File: src/render/renderer.cpp [hunk 1]
```diff
@@ -175,8 +175,23 @@ glm::vec4 Renderer::traceRayMIP(const Ray& ray, float sampleStep) const
...
-    return glm::vec4(isoColor, 1.0f);
+    glm::vec3 sample_pos = ray.origin + ray.tmin * ray.direction;
+    const glm::vec3 increment = sampleStep * ray.direction;
+    glm::vec4 color(0.0f);
+
+    for (float t = ray.tmin; t <= ray.tmax; t += sampleStep, sample_pos += increment) {
+        auto voxel_value = this->m_pVolume->getVoxelInterpolate(sample_pos);
+        if (voxel_value > this->m_config.isoValue) {
+            color = this->getTFValue(voxel_value);
+            // color = glm::vec4 { 0.8f, 0.8f, 0.2f, 1.0f };
```
Comment (anujinho) [inline src/render/renderer.cpp]: Final vals returned aren't the isoColor

**Review by anujinho — COMMENTED** [review]
done

File: src/render/renderer.cpp (hunk 1 above)
Comment (MihaiAnton) [reply src/render/renderer.cpp]: I know, but I think this will anyway be overwritten in the final version, right?

File: src/render/renderer.cpp [hunk 2]
```diff
@@ -167,31 +167,44 @@ glm::vec4 Renderer::traceRayMIP(const Ray& ray, float sampleStep) const
...
-        if (voxel_value > this->m_config.isoValue) {
-            color = this->getTFValue(voxel_value);
-            // color = glm::vec4 { 0.8f, 0.8f, 0.2f, 1.0f };
-            break;
+        for (float t = ray.tmin; t <= ray.tmax; t += sampleStep, sample_pos += increment) {
+            auto voxel_value = this->m_pVolume->getVoxelInterpolate(sample_pos);
+            if (voxel_value > this->m_config.isoValue) {
+                auto _color = glm::vec4 { 0.8f, 0.8f, 0.2f, 1.0f };
+                auto gradient = this->m_pGradientVolume->getGradientVoxel(sample_pos);
+                color = glm::vec4(this->computePhongShading(_color, gradient, this->m_pCamera->position(), this->m_pCamera->position()), 1.0f);
```
Comment (anujinho) [inline src/render/renderer.cpp]: why is the 'view' vector in the signature of phongshading the same as the light vector?

**Review by anujinho — COMMENTED** [review]
done

File: src/render/renderer.cpp (hunk 2 above)
Comment (MihaiAnton) [reply src/render/renderer.cpp]: I still left a TODO regarding this. I'm still not 100% sure how to get that parameter right.

**Pull request merged**